packages/clay-link/src/ClayLink.js

```diff
@@ -10,7 +10,30 @@ import templates from './ClayLink.soy.js';
  * Implementation of the Metal Clay Link.
  * @extends ClayComponent
  */
-class ClayLink extends ClayComponent {}
+class ClayLink extends ClayComponent {
+	/**
+	 * @inheritDoc
+	 */
+	attached() {
+		this.addListener('click', this._handleClick, true);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	disposed() {
+		this.removeListener('click', this._handleClick);
+	}
+
+	/**
+	 * Handles click.
+	 * @param {!Event} event
+	 * @protected
+	 */
+	_handleClick(event) {
+		this.emit('itemClicked', event);
+	}
+}
```

Comment:
@matuzalemsteles I'm not sure I understood correctly, but if by re-emitting the click event you mean something like this:

```js
	_handleClick(event) {
		this.emit('click', event);
	}
```

then I'm not sure, but it seems it would just create maximum call stack exceeded error, but we can give it a shot. This "itemClicked" event is also [pretty](https://github.com/liferay/clay/search?q=itemClicked&type=commits) common in the 2.x code base

Reply:
@julien Oh makes sense, you're right, it will cause an infinite loop. Well, I think we can keep it as is. I think we can encourage adding the `click` method to `defaultEventHandler` instead of `itemClicked` for having the same effect.

Reply:
> I think we can keep it as is. I think we can encourage adding the click method to defaultEventHandler instead of itemClicked for having the same effect.

@matuzalemsteles I don't understand what you mean by that. The way I see it, since this branch is there for maintenance, is to fix the issues using patterns that are already established. In this case I would leave this change as it is.

Reply:
@julien what I'm trying to say here is that the common default for other components like `ClayButton` in 2.x is to listen for `click` events instead of creating an event that already exists, I'm saying we have a bug when we don't use the means of listening to the event using the Metal.js defaults with `on` and `events` by Soy and JSX.

In the past, we had created the `defaultEventHandler` which looks similar to the `events` property but since it's not part of the Metal.js + Soy core, adding a method to the `defaultEventHandler` object will have no effect if it doesn't have any events being explicitly added the instance of the component so that Metal.js can monitor, for example:

```js
new metal.ClayButton(
  {
    label: 'Listening Click'
  },
  '#events-block'
)
.on('click', function(e) {
  alert('click');
});

// or
new metal.ClayButton(
  {
    label: 'Listening Click',
    events: {
      click: () => {},
    },
  },
  '#events-block'
);
```

This using in Soy:

```soy
{call ClayButton.render}
  {param events: [
    'click' : $_handleItemClick
  ] /}
{/call}
```

What I was explaining at the beginning is that not having any events explicitly added, whether calling the component or internally, will not invoke the `defaultEventHandler` method. Now since we have an explicit `emit` being added inside the component, the `defaultEventHandler` will be called, not necessarily just the method you want to invoke which is `itemClicked` but also `click`, so we can do that too.

```js
new metal.ClayLink(
	{
		defaultEventHandler: {
			click(event) {
				event.preventDefault();
			}
		},
		href: 'https://liferay.com',
	},
	'#link-icons'
);
```

In my last comment, I'm saying that we can use `click` now instead of using `itemClicked`.

Reply:
Yeah, in this case, I think we still need to emit an event for this to work, maybe just use `click` instead of `itemClicked` in `defaultEventHandler`. Just a matter of conversion.

Reply:
> Yeah, in this case, I think we still need to emit an event for this to work, maybe just use click instead of itemClicked in defaultEventHandler. Just a matter of conversion.

You lost me there ... if we emit click we're going to run in an "infinite loop", event if the `defaultEventHandler` object has a `click` method

Reply:
Additionally since most components use `"itemClicked"` why would we use `"click"`?

Reply:
> You lost me there ... if we emit click we're going to run in an "infinite loop", event if the defaultEventHandler object has a click method

Oh sorry, I'm not saying to re-emit the `click` event just next to the `defaultEventHandler` adding the `click` will work too. 

> Additionally since most components use "itemClicked" why would we use "click"?

Yes, I see some components with this, but I think it's more specific to components with an item structure, like a DropDown for example. Makes sense? For the common case for this, you use `click` like on the Button for example instead of adding `itemClicked`.

Reply:
> Oh sorry, I'm not saying to re-emit the click event just next to the defaultEventHandler adding the click will work too.

Sorry I still don't understand.
Without any change, this code 

```js
new metal.ClayLink(
	{
		defaultEventHandler: {
			click(event) {
				event.preventDefault();
			}
		},
		href: 'https://liferay.com',
	},
	'#link-icons'
);
```

won't do anything

Reply:
Yeah, exactly that doesn't work due to my explanation above https://github.com/liferay/clay/pull/4461#discussion_r761134837, but this will work after this implementation, I'm not saying to take this implementation away. you need to explicitly add a listener so that Metal.js can monitor and trigger the `defaultEventHandler`.

I'm implying that this is just a workaround to fix this problem of just adding a method to the `defaultEventHandler` that wouldn't trigger any method calls. And in that case, we can now add the `click` method that will work. Makes sense?

Reply:
> I'm implying that this is just a workaround to fix this problem of just adding a method to the defaultEventHandler that wouldn't trigger any method calls. And in that case, we can now add the click method that will work. Makes sense?

No sorry @matuzalemsteles I still don't understand what you mean.

Reply:
@matuzalemsteles I'm also confused by this discussion, not sure if you mean to add the onClick function like I first intended to?

Reply:
@georgel-pop-lr not sure either, let's see what @matuzalemsteles has to say.

Reply:
Well, I'll try to explain again, I'm not wanting to take this implementation away because it works as a workaround for the `defaultEventHandler` to work, see the ClayComponent component (https://github.com/liferay/clay/blob/2.x/packages/clay-component/src/ClayComponent.js) which edits some Metal.js internals so when someone issues an event it triggers the `defaultEventHandler` too.

What I'm saying is that with this implementation added in the PR of @georgel-pop-lr, we can also add the `click` method to `defaultEventHandler`, because adding a listener explicitly to the component causes it to the internals of Metal.js calls the `emit` method which is modified in ClayComponent, thus invokes the method in `defaultEventHandler`. So instead of adding the `handleItemClicked` to the `defaultEventHandler` on whoever is listening for the event, for example in the Demo, we can use `click` instead. Because that way we can keep the same consistency of who would do this with Soy or have the component instance by adding the event explicitly as I put in the example above.

Reply:
@matuzalemsteles I finally understood what you meant ... at least I think so.

What I'm seeing is that without any changes to the `ClayLink` component, this code can be used to handle click events.

```js
new metal.ClayLink(
	{
		ariaLabel: 'My Link',
		defaultEventHandler: {
			handleClick(event) {
				event.preventDefault();

				var linkElement = event.currentTarget;
				var href = linkElement.getAttribute('href');

				if (confirm('Are you sure you want to leave this site?')) {
					location.href = href;
				}
			}
		},
		href: 'https://liferay.com',
		icon: 'link',
		iconAlignment: 'right',
		label: 'Visit Liferay',
		spritemap: spritemap,
		title: 'Visit Liferay\'s website',
	},
	'#link-icons'
); 
```

Is that what you meant to say?
@georgel-pop-lr so basically what I understand is that only the markup changes would be needed (so we'll need to edit the pull request's title once again - sorry) and test the code example I pasted above.
If you need a hand, you know where to find me :wave:

Reply:
Hi all,

I have tested this solution with handleClick on 7.2.x, but it doesn't work.

Does this work in the demo files?

Regards.

Reply:
Well, not necessarily that we don't need these changes, more than adding an explicit event to the component, in the case like we're doing here, we can now `click` on the `defaultEventHandler` without any problems.

As I've commented in other comments, this wasn't working because adding a method to `defaultEventHandler` won't add any listeners to Metal.js event system, `defaultEventHandler` is fired only as a consequence of someone adding an explicit event using the common Metal.js methods, for example, the `events` property or the `on` method.

In the end, it is, with this implementation done by you, just use the `click` method on `defaultEventHandler` now.